proposals/0021-vk-coop-matrix.md

```diff
@@ -0,0 +1,264 @@
+(SPV_KHR_cooperative_matrix)[https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_cooperative_matrix.html]
+extension.
+
+To help implement `vk::khr::CooperativeMatrix`, we will also introduce a utility
+class `vk::util::ArithmeticSelector`. This class will be defined in a header
+file "vk/util/arithmetic_selector.h". It can be used to generate inline Spir-V
+arithmetic instructions with the correct opcode for the type.
+
+The interface for `vk::khr::CooperativeMatrix` uses enums that are defined in
+the Spir-V specification. The required enums will be defined in "vk/spirv.h".
+
+## Detailed design
+
+### `vk::khr::CooperativeMatrix`
+
+The `CooperativeMatrix` class will defined in a header file as a wrapper around
+a `vk::SpirvType` that will expand to the appropriate SPIR-V type. This class
+will have the following interface.
+
+```c++
+
+// The base cooperative matrix class. The template arguments correspond to the
+// operands in the OpTypeCooperativeMatrixKHR instruction.
+template <typename ComponentType, Scope scope, uint rows, uint columns,
+CooperativeMatrixUse use>
+class CooperativeMatrix {
+
+// Apply OpSNegate or OFNegate, depending on ComponentType, in a element by element manner.
+CooperativeMatrix negate();
+
+// Apply OpIAdd or OFAdd, depending on ComponentType, in a element by element manner.
+CooperativeMatrix operator+(CooperativeMatrix other);
+
+// Apply OpISub or OFSub, depending on ComponentType, in a element by element manner.
+CooperativeMatrix operator-(CooperativeMatrix other);
+
+// Apply OpIMul or OFMul, depending on ComponentType, in a element by element manner.
+CooperativeMatrix operator*(CooperativeMatrix other);
+
+// Apply OpSDiv, OpUDiv or OFDiv, depending on ComponentType, in a element by element manner.
+CooperativeMatrix operator/(CooperativeMatrix other);
+
+// Apply OpMatrixTimesScalar in a element by element manner.
+CooperativeMatrix operator*(ComponentType scalar);
+
+// Store the cooperative matrix using OpCooperativeMatrixStoreKHR to data[i] using
+// memory layout RowMajorKHR.￼
+void StoreRowMajor(RWStructuredBuffer<ComponentType> data, uint32_t index);
+
+// Store the cooperative matrix using OpCooperativeMatrixStoreKHR to data[i] using
+// memory layout ColumnMajorKHR.￼
+void StoreColumnMajor(RWStructuredBuffer<ComponentType> data, uint32_t index);
+
+// Store the cooperative matrix using OpCooperativeMatrixStoreKHR to data[i] using
+// memory layout RowMajorKHR and the given stride and memory access mask.
+void StoreRowMajor(RWStructuredBuffer<ComponentType> data, uint32_t index,
+uint32_t stride, MemoryAccessMask memoryAccessMask);
+
+// Store the cooperative matrix using OpCooperativeMatrixStoreKHR to data[i] using
+// memory layout ColumnMajorKHR and the given stride and memory access mask.
+void StoreColumnMajor(RWStructuredBuffer<ComponentType> data, uint32_t index,
+uint32_t stride, MemoryAccessMask memoryAccessMask);
```

Comment:
These need to be extended to be more like the loads. In particular, we want to be able to store to an array in groupshared memory.

Reply:
Still working on how to allow groupshared memory. I don't see how to do with without allowing more that is legal. In this case, spirv-val will be the backstop, but that is not ideal.

There are also some issues with spirv-opt not optimizing correctly when the groupshared memory is used. I'll need to get another fix into spirv-opt.

Reply:
We cannot have a wrapper function that will load and store from group shared memory because HLSL does not allow us to pass that type of memory by reference. If someone want to do a loads and stores they will have to access the ext_instruction and SPIR-V types directly.

Reply:
Adding pointers/references to the type system is kinda a pain, but I wonder if we could do something slightly hacky that doesn't actually change the language to address cases like this.

We do have the ability to represent `ref` passed parameters in DXC's intrinsics, so maybe we could have a builtin that converts a `ref groupshared T` to a `GroupsharedAddr<T>` builtin object, that has no methods and can't be used for anything other than passing into some limited set of builtin functions.

Thoughts?

Reply:
Interesting idea. I started thinking about this, and I realized we can do something similar using only inline spir-v. See https://godbolt.org/z/WecY1a9K5 for how we can define an opaque pointer type that can be passed around, but only really used in inline spir-v defined function and builtin functions.

Reply:
I added an opaque pointer type, and a function to get the address of the groupshared memory.